### PR TITLE
Iris wipe transition modal page route

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
This wasn't what I set out to implement, but I submit the PRs I end up with, not the ones I wish I had.

**Comment by Hixie** [comment]
@abarth

**Comment by Hixie** [comment]
(There's a comment in this PR about an IrisWipeTransition class. That'll come in a subsequent patch.)

**Comment by Hixie** [comment]
The performance of this transition is pathetic, but I suspect it's the same issue as the FAB animation. It's all in the GPU. I'm building three new circle shadows on every frame. I suspect Skia is trying to oversample the shadows again.

**Comment by abarth** [comment]
Yeah, circular shadows are slow in skia because of https://bugs.chromium.org/p/skia/issues/detail?id=5224

Maybe we should hold off on the shadows until that issue is fixed?

File: examples/material_gallery/lib/demo/grid_list_demo.dart [hunk 1]
```diff
@@ -54,21 +54,38 @@ class GridDemoPhotoItem extends StatelessWidget {
...
 
-    Navigator.push(context, new MaterialPageRoute<Null>(
+    RenderBox photoBox = context.findRenderObject();
+    RenderBox navigatorBox = ModalRoute.of(context).navigator.context.findRenderObject();
+    Size navigatorSize = navigatorBox.size;
+    Point pixelCenter = navigatorBox.globalToLocal(photoBox.localToGlobal(photoBox.size.center(Point.origin)));
+    FractionalOffset fractionalCenter = new FractionalOffset(
+      pixelCenter.x / navigatorSize.width,
+      pixelCenter.y / navigatorSize.height
+    );
```
Comment (abarth) [inline examples/material_gallery/lib/demo/grid_list_demo.dart]: O_o

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 2]
```diff
@@ -904,6 +904,18 @@ class FractionalOffset {
     return new Point(rect.left + dx * rect.width, rect.top + dy * rect.height);
   }
 
+  double distanceToNearestCorner(Size size) {
+    final double x = (dx <= 0.5 ? dx : 1.0 - dx) * size.width;
+    final double y = (dy <= 0.5 ? dy : 1.0 - dy) * size.height;
+    return math.sqrt(x * x + y * y);
+  }
+
+  double distanceToFurthestCorner(Size size) {
```
Comment (abarth) [inline packages/flutter/lib/src/painting/box_painter.dart]: Farthest

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 3]
```diff
@@ -845,25 +845,29 @@ abstract class _RenderCustomClip<T> extends RenderProxyBox {
...
   }
 
   T get _defaultClip;
-  T _clip;
+
+  T _cachedClip;
+  T get _clip {
+    assert(!needsLayout);
+    if (_cachedClip == null)
+      _cachedClip = _clipper?.getClip(size) ?? _defaultClip;
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: `?=`

**Comment by abarth** [comment]
The setup is kind of crazytown.  Is there a path to making that better?

Also, please add a test.  For example, you could clock the animation halfway and then show that you can't tap on something that's being clipped.

LGTM

File: examples/material_gallery/lib/demo/grid_list_demo.dart (hunk 1 above)
Comment (Hixie) [inline examples/material_gallery/lib/demo/grid_list_demo.dart]: Yeah I wasn't really sure how else to do it. We keep getting this problem where we want one thing to depend on the (layout) location of another in the same build.

**Comment by Hixie** [comment]
Upon discussion we decided to change this so that instead the hero controls where the circle's center is. This reduces the "where to center the circle" craziness to a previously unsolved problem, but at least that means we only have to solve the problem once...

**Comment by Hixie** [comment]
Closing for now.

**Pull request closed without merging**